CSC8503/InventoryBuffSystem/PlayerInventory.cpp

```diff
@@ -1,95 +1,96 @@
-				{
-					mOnItemDroppedFunctionMap[inItem](playerNo);
-				}
-
-				mPlayerInventory[playerNo][invSlot] = none;
+				Notify(mOnItemDroppedInventoryEventMap[inItem],playerNo);
 			}
-		}
-	}
-
-	void PlayerInventory::DropItemFromPlayer(int playerNo, int invSlot)
-	{
-		CreateItemPickup(mPlayerInventory[playerNo][invSlot],
-			(mPlayerGameObjectsPTR + playerNo)->GetTransform().GetPosition());
 
-		if (mOnItemDroppedFunctionMap.find(mPlayerInventory[playerNo][invSlot]) != mOnItemDroppedFunctionMap.end())
-		{
-			mOnItemDroppedFunctionMap[mPlayerInventory[playerNo][invSlot]](playerNo);
+			mPlayerInventory[playerNo][invSlot] = none;
 		}
-
-		mPlayerInventory[playerNo][invSlot] = none;
 	}
+}
 
-
-	void PlayerInventory::DropFlagFromPlayer(int playerNo)
+void PlayerInventory::DropItemFromPlayer(int playerNo, int invSlot)
+{
+	if (mOnItemDroppedInventoryEventMap.find(mPlayerInventory[playerNo][invSlot]) != mOnItemDroppedInventoryEventMap.end())
 	{
-		for (int invSlot = 0; invSlot < MAX_INVENTORY_SLOTS; invSlot++)
-		{
-			if (mPlayerInventory[playerNo][invSlot] == flag)
-			{
-				CreateItemPickup(flag, *mFlagLocationPTR);
-				mPlayerInventory[playerNo][invSlot] = none;
-			}
-		}
+		Notify(mOnItemDroppedInventoryEventMap[mPlayerInventory[playerNo][invSlot]],playerNo);
 	}
 
-	void PlayerInventory::UseItemInPlayerSlot(int playerNo, int invSlot)
-	{
-		if (mOnItemUsedFunctionMap.find(mPlayerInventory[playerNo][invSlot]) != mOnItemUsedFunctionMap.end())
-		{
-			mOnItemUsedFunctionMap[mPlayerInventory[playerNo][invSlot]](playerNo);
-		}
+	mPlayerInventory[playerNo][invSlot] = none;
+}
 
-		mPlayerInventory[playerNo][invSlot] = none;
+void PlayerInventory::UseItemInPlayerSlot(int playerNo, int invSlot)
+{
+	if (mOnItemUsedInventoryEventMap.find(mPlayerInventory[playerNo][invSlot]) != mOnItemUsedInventoryEventMap.end())
+	{
+		Notify(mOnItemUsedInventoryEventMap[mPlayerInventory[playerNo][invSlot]],(playerNo));
 	}
 
-	PlayerInventory::item PlayerInventory::GetRandomItemFromPool(unsigned int seed)
-	{
-		std::mt19937 rng(seed);
-		std::shuffle(mItemsInRandomPool.begin(), mItemsInRandomPool.end(), rng);
-		return mItemsInRandomPool[0];
+	mPlayerInventory[playerNo][invSlot] = none;
+}
+
+void InventoryBuffSystem::PlayerInventory::Attach(PlayerInventoryObserver* observer)
+{
+	mInventoryObserverList.push_back(observer);
+}
+
+void InventoryBuffSystem::PlayerInventory::Detach(PlayerInventoryObserver* observer)
+{
+	mInventoryObserverList.remove(observer);
+}
+
+void InventoryBuffSystem::PlayerInventory::Notify(const InventoryEvent invEvent,int playerNo)
+{
+	std::list<PlayerInventoryObserver*>::iterator iterator = mInventoryObserverList.begin();
+	while (iterator != mInventoryObserverList.end()) {
+		(*iterator)->UpdateInventoryObserver(invEvent,playerNo);
+		++iterator;
 	}
+}
+
+PlayerInventory::item PlayerInventory::GetRandomItemFromPool(unsigned int seed)
+{
+	std::mt19937 rng(seed);
```

Comment:
You may need to coordinate with Eren with this: random number seeds must be shared with all players in multiplayer, otherwise we may end up with different random numbers floating about.